packages/data-api/src/AnalyticsFetchQuery.js

```diff
@@ -3,124 +3,172 @@
+  );
+};
+
+export class AnalyticsFetchQuery {
   constructor(database, options) {
-    super(database, options);
+    this.database = database;
+
+    const { dataElementCodes, organisationUnitCodes, startDate, endDate } = options;
+    this.dataElementCodes = dataElementCodes;
+    this.entityCodes = organisationUnitCodes;
+    this.startDate = startDate;
+    this.endDate = endDate;
+
+    this.aggregations = [];
+    this.originalAggregationsProcessed = [];
+    if (options.aggregations && options.canProcessAggregations) {
+      for (let i = 0; i < options.aggregations.length; i++) {
+        const aggregation = options.aggregations[i];
+        const aggregationSwitch = AGGREGATION_SWITCHES[aggregation?.type];
+        if (!aggregationSwitch || !supportsConfig(aggregationSwitch, aggregation?.config)) {
+          break; // We only support chaining aggregations up to the last supported type
+        }
+        const dbAggregation = { type: aggregation?.type };
+        dbAggregation.switch = AGGREGATION_SWITCHES[aggregation?.type]; // add internal switch
+        dbAggregation.config = aggregation?.config; // add external config, supplied by client
+        dbAggregation.stackId = i + 1;
+        this.aggregations.push(dbAggregation);
+        this.originalAggregationsProcessed.push(aggregation);
+      }
+    }
 
-    const firstAggregation = options.aggregations?.[0];
-    this.aggregation = { type: firstAggregation?.type };
-    this.aggregation.switches = AGGREGATION_SWITCHES[firstAggregation?.type]; // add internal switches
-    this.aggregation.config = firstAggregation?.config; // add external config, supplied by client
-    this.isAggregating = !!this.aggregation.switches;
+    this.isAggregating = this.aggregations.length > 0;
 
     this.validate();
   }
 
-  validate() {
-    if (this.aggregation.switches?.aggregateEntities && !this.aggregation.config?.orgUnitMap) {
-      throw new Error('When using entity aggregation you must provide an org unit map');
-    }
+  async fetch() {
+    const { query, params } = this.buildQueryAndParams();
+
+    const sqlQuery = new SqlQuery(query, params);
+
+    return {
+      analytics: await sqlQuery.executeOnDatabase(this.database),
+      aggregationsProcessed: this.originalAggregationsProcessed,
+    };
   }
 
-  getEntityCodeField() {
-    return this.aggregation.switches?.aggregateEntities ? 'aggregation_entity_code' : 'entity_code';
+  validate() {
+    this.aggregations.forEach(aggregation => {
+      if (aggregation.switch.aggregateEntities && !aggregation.config?.orgUnitMap) {
+        throw new Error('When using entity aggregation you must provide an org unit map');
+      }
+    });
   }
 
-  getCommonFields() {
-    return [this.getEntityCodeField(), 'data_element_code'];
+  getEntityCodeField(aggregation) {
+    return aggregation.switch.aggregateEntities ? 'aggregation_entity_code' : 'entity_code';
   }
 
-  getEntityCommonTableExpression() {
+  getCtesAndParams(aggregation) {
```

Comment:
Rather than keeping the params in some internal state, I moved these functions to be more of 'build a query part and returns the associated params' kind of thing. I did that because the order of the params is very important, and so when building up query parts, the caller must ensure that the parts and the params are inserted in the same order.